pep-0558.rst

```diff
@@ -262,54 +285,66 @@ defined ``locals()`` builtin, trace functions necessarily use the implementation
 dependent ``frame.f_locals`` interface, as a frame reference is what gets
 passed to hook implementations.
 
-Instead of being a direct reference to the dynamic snapshot returned by
-``locals()``, ``frame.f_locals`` will be updated to instead return a dedicated
-proxy type (implemented as a private subclass of the existing
-``types.MappingProxyType``) that has two internal attributes not exposed as
-part of either the Python or public C API:
-
-* *mapping*: the dynamic snapshot that is returned by the ``locals()`` builtin
-* *frame*: the underlying frame that the snapshot is for
+Instead of being a direct reference to the dynamic snapshot historically
+returned by ``locals()``, ``frame.f_locals`` will be updated to instead return
+a dedicated fast locals proxy type that has an internal reference back to the
+underlying frame object. As with the existing function level ``f_locals``
+attribute, this proxy object will only be created on demand, rather than being
+present by default.
 
-``__setitem__`` and ``__delitem__`` operations on the proxy will affect not only
-the dynamic snapshot, but *also* the corresponding fast local or cell reference
-on the underlying frame.
+``__getitem__`` operations on the proxy will read the corresponding fast local
+or cell reference on the underlying frame rather than storing a separate
+mapping from keys to values.
 
-The ``locals()`` builtin will be made aware of this proxy type, and continue to
-return a reference to the dynamic snapshot rather than to the write-through
-proxy.
+Similarly, ``__setitem__`` and ``__delitem__`` operations on the proxy will
+directly affect the corresponding fast local or cell references on the
+underlying frame.
 
-At the C API layer, ``PyEval_GetLocals()`` will implement the same semantics
-as the Python level ``locals()`` builtin, and a new
-``PyFrame_GetPyLocals(frame)`` accessor API will be provided to allow the
-function level proxy bypass logic to be encapsulated entirely inside the frame
-implementation.
+Attempting to set or delete a key through the proxy that isn't defined as a
+local variable or closure reference on the underlying frame will raise
+``KeyError``. Attempting to delete a key that has already been deleted from
```

Comment:
Unfortunately, Trio depends on being able to write to unrecognized keys in `f_locals`: https://github.com/python-trio/trio/blob/1e95ee42ddf34efd58b94346edbd7a15f6f0f094/trio/_core/_run.py#L1142-L1144

It's a terrible hack and we really should switch to something better, but I think right now there is nothing better – we need a way to mark frames that's visible to a signal handler, and `f_locals` is the only option. So I think we'll have to provide some alternative before we can stop stashing random data in `f_locals`.

Are you aiming for 3.8 or 3.9? If you're aiming for 3.9 then we might be able to do both changes at the same time, so you never have to write the code for handling arbitrary keys in the new `f_locals` proxy object.

Reply:
At the moment, I'm still aiming for 3.8, and dropping the mapping turned out not to be as clear a win at the implementation level as I thought it would be, since it would mean the mutable fast locals proxy can't readily share code with the immutable mapping proxy any more.

So as things stand, I actually have the code to handle arbitrary keys already written - there is just a related bug in the frame clearing logic, where it isn't tidying up those references properly, so some objects are being kept alive when they shouldn't be.

So I'll revert the frame related PEP changes from this PR, and focus on fixing that reference clearing bug in the existing implementation.

Reply:
Latest PR update has restored support for arbitrary keys in `frame.f_locals`.